pontoon/static/js/utils/data.js

```diff
@@ -0,0 +1,85 @@
+
+import React from 'react';
+
+import {ajax} from 'utils/ajax';
+
+
+export class DataManager {
+
+    constructor (state) {
+        this.state = state;
+    }
+
+    get data () {
+        return this.state.data;
+    }
+
+    get errors () {
+        return this.state.errors;
+    }
+}
+
+
+export function dataManager(WrappedComponent, Manager, data) {
+
+    return class Wrapper extends React.Component {
+
+        constructor(props) {
+            super(props);
+            this.state = {errors: {}, data: data};
+        }
+
+        get api () {
+            return this.props.api;
+        }
+
+        get requestMethod () {
+            return this.props.requestMethod || 'get';
+        }
+
+        get submitMethod () {
+            return this.props.submitMethod || this.requestMethod;
+        }
+
+        refreshData = async (params) => {
+            return this.handleResponse(
+                await ajax.fetch(
+                    this.api,
+                    params,
+                    {method: this.requestMethod}));
+        }
+
+        handleResponse = async (response) => {
+            const {status} = response;
+            const json = await response.json();
+            if (status === 200) {
+                const {data} = json;
+                this.setState({data, errors: {}});
+                return;
+            }
+            const {errors} = json;
+            this.setState({errors});
+        }
+
+        handleSubmit = async (params) => {
+            return this.handleResponse(
+                await ajax.fetch(
+                    this.api,
+                    params,
+                    {method: this.requestMethod}));
+        }
+
+        render() {
+            Manager = Manager || DataManager;
+            const manager = new Manager(this.state);
+            return (
+                <WrappedComponent
+                   manager={manager}
+                   errors={manager.errors}
+                   data={manager.data}
+                   handleSubmit={this.handleSubmit}
+                   refreshData={this.refreshData}
+                   {...this.props} />);
+        }
+    };
+}
```

Comment:
The content of this file is too complex, there are too many concepts and it is not obvious from the outside what this does, or how one uses it. I for one know that I won't use it because I'm sure next time I look at this I will have forgotten how it works. I'm not even 100% sure what problem you're solving with this HOC. In short, I don't think this follows the KISS principle -- it's not simple, and definitely not stupid.

Reply:
im not sure how i could have done it better - but hopefully time and usage will show us

there are some things that are really common patterns and in react-land HOCs are the way to factor that out - ive tried to stay by the book in what we eventually landed this (even tho it meant you cant do what i had hoped to at a higher level of abstraction. /endmoan)

what is not really addressed/factored into this code is how pages are reloaded with partial content <> full page loads. But my mission is `string_tiers` so im really reluctant to think through those issues in this context.

either way - i think we have something usable - if not, its something to crititque/rubbish/remove/replace - but in the meantime we can use it 8/

Reply:
> what is not really addressed/factored into this code is how pages are reloaded with partial content <> full page loads

@adngdb on this point i think it should `Object.assign(state.data, response.data)`

Reply:
I'm missing the context so forgive my ignorance but since I read the code snippet above, please note [state should be treated as if it was immutable](https://reactjs.org/docs/react-component.html#state):

> Never mutate `this.state` directly, as calling `setState()` afterwards may replace the mutation you made. Treat `this.state` as if it were immutable.

So something like `Object.assign({}, state.data, response.data)` is preferable, and then `setState()` as needed.

Reply:
hi - sorry that was more meant to be pseudo code for what should happen, rather than the implementation

Reply:
Cool, the code drew my attention — sorry for the noise!

Reply:
@phlax I don't follow, what do you mean by
> what is not really addressed/factored into this code is how pages are reloaded with partial content <> full page loads

?

Reply:
if you make a json call back - eg for partial content for a dashboard - you dont want to cane the existing `state` object more add to/update it

Reply:
to (hopefully) clarify a bit better - here is the implementation/change i had in mind:

```
@@ -54,7 +58,7 @@ export function dataManager(WrappedComponent, Manager, data) {
             const json = await response.json();
             if (status === 200) {
                 const {data} = json;
-                this.setState({data, errors: {}});
+                this.setState(prevState => ({data: Object.assign(prevState.data, {...data}), errors: {}}));
                 return;
             }
```

Reply:
That seems counter intuitive to me. Wouldn't the service you're calling send all the data you need? Maybe it can send back an object that doesn't have a key the previous state had, for good reason. In such a case, your state would still keep the obsolete key.

Reply:
i can show you some code, just not yet - but i think this change does make sense in the context of sometimes retrieving all data and sometimes only updating the existing